cpp/src/arrow/extension/fixed_shape_tensor.cc

```diff
@@ -0,0 +1,323 @@
+    }
+    case Type::UINT16: {
+      value_array = std::make_shared<UInt16Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT16: {
+      value_array = std::make_shared<Int16Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::UINT32: {
+      value_array = std::make_shared<UInt32Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT32: {
+      value_array = std::make_shared<Int32Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::UINT64: {
+      value_array = std::make_shared<Int64Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::INT64: {
+      value_array = std::make_shared<Int64Array>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::HALF_FLOAT: {
+      value_array = std::make_shared<HalfFloatArray>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::FLOAT: {
+      value_array = std::make_shared<FloatArray>(tensor->size(), tensor->data());
+      break;
+    }
+    case Type::DOUBLE: {
+      value_array = std::make_shared<DoubleArray>(tensor->size(), tensor->data());
+      break;
+    }
+    default: {
+      return Status::NotImplemented("Unsupported tensor type: ",
+                                    tensor->type()->ToString());
+    }
+  }
+  arr = std::make_shared<FixedSizeListArray>(ext_type->storage_type(), tensor->shape()[0],
+                                             value_array);
+  auto ext_data = arr->data();
+  ext_data->type = ext_type;
+  return ext_type->MakeArray(ext_data);
+}
+
+Result<std::shared_ptr<Tensor>> FixedShapeTensorType::ToTensor(
+    std::shared_ptr<Array> arr) {
+  // To convert an array of n dimensional tensors to a n+1 dimensional tensor we
+  // interpret the array's length as the first dimension the new tensor. Further, we
+  // define n+1 dimensional tensor's strides by front appending a new stride to the n
+  // dimensional tensor's strides.
+
+  ARROW_RETURN_IF(arr->null_count() > 0,
+                  Status::Invalid("Null values not supported in tensors."));
+
+  auto ext_arr = internal::checked_pointer_cast<FixedSizeListArray>(
+      internal::checked_pointer_cast<ExtensionArray>(arr)->storage());
+  ARROW_RETURN_IF(!is_fixed_width(ext_arr->value_type()->id()),
+                  Status::Invalid(ext_arr->value_type()->ToString(),
+                                  " is not valid data type for a tensor"));
+
+  std::vector<int64_t> shape = this->shape();
+  shape.insert(shape.begin(), 1, arr->length());
+
+  int64_t major_stride;
+  std::vector<int64_t> tensor_strides = this->strides();
+  if (internal::MultiplyWithOverflow(ext_arr->value_type()->byte_width(),
+                                     ext_arr->list_type()->list_size(), &major_stride)) {
+    return Status::Invalid("Overflow in tensor strides");
+  }
+  if (!this->permutation().empty()) {
+    internal::Permute(this->permutation(), &tensor_strides);
```

Comment:
If you permute the strides, shouldn't you also permute the shape? (or the `this->strides()` were already permuted, so you are permuting "back" here?)

Reply:
`FixedShapeTensorType::shape()` now permutes `shape_` before returning.